File: javascript/test/instance/GenericCollectionHolder_FindFirstOrNullAlias.ts
//··························································
// Copyright (c) 2023-2025. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {NullOr}                                     from "@joookiwi/type"
import type {BooleanCallback, RestrainedBooleanCallback} from "../../src/type/callback"
import type {StraightCollectionHolderForTest}            from "./StraightCollectionHolderForTest"

import {GenericCollectionHolder} from "../../src/GenericCollectionHolder"
import {ABCD}                    from "../value/arrays"

export class GenericCollectionHolder_FindFirstOrNullAlias
    extends GenericCollectionHolder<string>
    implements StraightCollectionHolderForTest<string> {

    public amountOfCall = 0

    public constructor() { super(ABCD,) }

    public execute(action: (instance: this,) => void,): this {
        action(this,)
        return this
    }

    public override findFirstOrNull<const S extends string, >(predicate: RestrainedBooleanCallback<string, S>,): NullOr<S>
    public override findFirstOrNull(predicate: BooleanCallback<string>,): NullOr<string>
    public override findFirstOrNull(predicate: BooleanCallback<string>,) {
        this.amountOfCall++
        return super.findFirstOrNull(predicate,)
    }

}

File: java/collection.joookiwi/src/test/java/instance/GenericCollectionHolder_FindFirstOrNullAlias.java
package instance;

import java.util.function.BooleanSupplier;
import java.util.function.Consumer;
import java.util.function.Predicate;
import joookiwi.collection.java.GenericCollectionHolder;
import joookiwi.collection.java.callback.ObjIntPredicate;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import static joookiwi.collection.java.CommonContracts.ALWAYS_THIS_1;
import static joookiwi.collection.java.NumericConstants.MAX_INT_VALUE;
import static value.Arrays.ABCD;

@NotNullByDefault
public final class GenericCollectionHolder_FindFirstOrNullAlias
        extends GenericCollectionHolder<String>
        implements StraightCollectionHolderForTest<String, GenericCollectionHolder_FindFirstOrNullAlias> {

    /// The field that tell the amount of time the method `findFirstOrNull` has been called
    public @Range(from = 0, to = MAX_INT_VALUE) int amountOfCall = 0;

    public GenericCollectionHolder_FindFirstOrNullAlias() { super(ABCD); }

    @Override public @Range(from = 0, to = MAX_INT_VALUE) int getAmountOfCall() { return amountOfCall; }

    @Contract(ALWAYS_THIS_1)
    @Override public GenericCollectionHolder_FindFirstOrNullAlias execute(final Consumer<GenericCollectionHolder_FindFirstOrNullAlias> action) {
        action.accept(this);
        return this;
    }


    @Override public @Nullable String findFirstOrNull(final ObjIntPredicate<? super String> predicate) {
        amountOfCall++;
        return super.findFirstOrNull(predicate);
    }

    @Override public @Nullable String findFirstOrNull(final Predicate<? super String> predicate) {
        amountOfCall++;
        return super.findFirstOrNull(predicate);
    }

    @Override public @Nullable String findFirstOrNull(final BooleanSupplier predicate) {
        amountOfCall++;
        return super.findFirstOrNull(predicate);
    }

}
